packages/SwingSet/docs/vat-environment.md

```diff
@@ -0,0 +1,113 @@
+
+Browser environments also have a huge list of [other features](https://developer.mozilla.org/en-US/docs/Web/API) which are presented as names in the global scope (some of which were also added to Node.js). None of these are available in a SES environment. The most surprising removals include `atob`, `TextEncoder`, and `URL`.
+
+`debugger` is a first-class JavaScript statement, and behaves as expected in vat code.
+
+## Other Changes
+
+The [shim](https://github.com/Agoric/SES-shim/) which provides our SES environment is not as fully-featured as a native implemenation could be. As a result, some forms of code cannot be used yet. These restrictions should be lifted once the JS engine you use can provide SES natively:
+
+### HTML Comments
+
+JavaScript parsers may or may not recognize HTML comments within source code, potentially leading to different behavior on different engines. For safety, the SES shim rejects any source code which looks like it might contain a comment open (`<!--`) or close (`-->`) sequence. The filter uses a regular expression, not a full parser.
+
+### Dynamic Import Expressions
+
+One active JS feature proposal would add a "dynamic import" expression: `await import('path')`. If the engine implements this, vat code might be able to bypass the Compartment's module map. For safety, the SES shim rejects code that looks like it uses this feature. The regular expression that looks for this pattern can be confused into falsely rejecting legitimate code, usually by having the word `import` in a comment at the end of a line.
+
+### Direct Eval Expressions
+
+A "direct eval", invoked like `eval(code)`, is defined to behave as if `code` were expanded in place. The evaluated code sees the same scope as the `eval` itself sees, so the `code` in:
+
+```js
+function foo(code) {
+  const x = 1;
+  eval(code);
+}
+```
+
+gets to reference `x`. If you perform a direct eval, you cannot hide your internal authorities from the code you're evaluating.
+
+In contrast, an "indirect eval" only gets the global scope, not the local scope. In a safe SES environment, indirect eval is a useful and common tool. The evaluated code can only access global objects, and those are all safe (and frozen). The only bad thing an indirect eval can do is consume unbounded CPU or memory. Once you've evaluated that code, you can give it as many or as few authories as you like, by invoking it with arguments.
+
+The most common way to invoke an indirect eval is like `(1,eval)(code)`.
+
+The SES shim cannot correctly emulate a "direct eval". If it tried, it would wind up performing an indirect eval. This could be pretty confusing, because the code might not actually use objects from the local scope, so you might not notice the problem until some future change altered the behavior.
```

Comment:
Remove quotes on "direct eval".

"would wind up performing" -> "it would perform"

"scope, so" -> "scope. You might..."

"future change" -> "later change"